### PR TITLE
refactor(logging): move activity/flow event decision out of log object

### LINKED ISSUE #1284: Refactor metricsContext stuff to stop passing so many objects around
From https://github.com/mozilla/fxa-auth-server/pull/1259/files#r65112250:

> We're doing a lot of passing objects around here, passing in a `metricsContext` to each route and them passing the `request` to a bunch of logging calls.  I wonder if we could clean this up by changing around our abstractions.  What if we have some Hapi plugin (or whatever it's called) that added a metrics object to the incoming request, so we could do something like:
>
> ```
> return request.app.metrics.event('delete', { ...data... });
> return request.app.metrics.activityEvent('account.signed', { ...data... });
> ```
>
> Do you think something like that might come out cleaner overall?

### PR DESCRIPTION
**Pull request by philbooth**
Part 3/3, fixes #1284. Part 1 was #1500 and part 2 was #1503.

Hopefully this change resolves the recent confusion around what is an activity event and what is a flow event. From now on, we need only make calls to `request.emitEvent` for both event types. All of the decision making for which `log` method to call is contained therein.

This also makes `lib/log.js` a little bit cleaner because it enabled me to move the decision-making into a dedicated module, `lib/events.js`. So `log.activityEvent` no longer includes a conditional call to `log.flowEvent`, those two methods are properly separated.

The new module takes the approach of defining all of the events that *aren't* a flow event. This is easier than defining the flow events directly because of how many there will be when we're finished. The body of the new method just checks the event name against these definitions before calling out to `log.activityEvent` and/or `log.flowEvent`.

Here are the salient log lines from a sample Sync sign-in:

```
activityEvent {"event":"account.login","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","service":"sync","uid":"d761ffbe7e134eecba53037f101ffa3d"}
flowEvent {"event":"account.login","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","time":1476796184340,"flow_id":"7deab7154fd0c5b0d81ab2256f7c0fdca5b51b3bc93c2532dc083a3880594d3a","flow_time":8139,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
activityEvent {"event":"account.confirmed","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","service":"sync","uid":"d761ffbe7e134eecba53037f101ffa3d"}
flowEvent {"event":"account.confirmed","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","time":1476796222371,"flow_id":"7deab7154fd0c5b0d81ab2256f7c0fdca5b51b3bc93c2532dc083a3880594d3a","flow_time":46170,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
activityEvent {"event":"account.keyfetch","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","uid":"d761ffbe7e134eecba53037f101ffa3d"}
flowEvent {"event":"account.keyfetch","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","time":1476796229593,"flow_id":"7deab7154fd0c5b0d81ab2256f7c0fdca5b51b3bc93c2532dc083a3880594d3a","flow_time":53392,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
activityEvent {"event":"device.created","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","uid":"d761ffbe7e134eecba53037f101ffa3d","device_id":"3c1a0a03e87e64c02c73f20552349ed2","is_placeholder":true}
activityEvent {"event":"account.signed","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","uid":"d761ffbe7e134eecba53037f101ffa3d","account_created_at":1476796184346,"device_id":"3c1a0a03e87e64c02c73f20552349ed2"}
flowEvent {"event":"account.signed","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:52.0) Gecko/20100101 Firefox/52.0","time":1476796229632,"flow_id":"7deab7154fd0c5b0d81ab2256f7c0fdca5b51b3bc93c2532dc083a3880594d3a","flow_time":53431,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
```

@vbudhram r?

### CODE REVIEW

File: lib/events.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  'device.deleted': true,
+  'device.updated': true
+}
+
+const NOT_FLOW_EVENTS_IF_SERVICE_IS_CONTENT_SERVER = {
+  'account.signed': true
+}
+
+module.exports = log => {
+  return { emit: emit }
```
Comment (seanmonstar) [inline lib/events.js]: We can es6 harder than that!

```js
return {
  emit(event, data) {
    // ...
  }
}
```

File: lib/events.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        return log.activityEvent(event, request, data)
+      }
+    })
+    .then(() => {
+      if (NOT_FLOW_EVENTS[event]) {
+        return
+      }
+
+      if (
+        NOT_FLOW_EVENTS_IF_SERVICE_IS_CONTENT_SERVER[event] &&
```
Comment (seanmonstar) [inline lib/events.js]: Dat name doe...

File: lib/events.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const P = require('./promise')
+
+const ACTIVITY_EVENTS = {
```
Comment (seanmonstar) [inline lib/events.js]: Have you thought much about `Set` that is in ES6?

```js
const EVENTS = new Set([
  'a',
  'b',
  'c'
])

assert(EVENTS.has('a'))
assert(!EVENTS.has('toString'))
```

File: lib/server.js [hunk 4]
```diff
@@ -297,11 +297,13 @@ function create(log, error, config, routes, db) {
   )
 
   const metricsContext = require('./metrics/context')(log, config)
-
   server.decorate('request', 'stashMetricsContext', metricsContext.stash)
   server.decorate('request', 'gatherMetricsContext', metricsContext.gather)
   server.decorate('request', 'validateMetricsContext', metricsContext.validate)
 
+  const events = require('./events')(log)
+  server.decorate('request', 'emitEvent', events.emit)
```
Comment (seanmonstar) [inline lib/server.js]: I like it.

File: lib/events.js (hunk 2 above)
Comment (philbooth) [inline lib/events.js]: What specifically don't you like about it? Just the length or something else?

Fwiw, I wanted something named to replace the magic string that we currently use in `log.flowEvent`. And I wanted the name to be explicit about what it's for.

If you can suggest something more suitable, I'm happy to change it.

File: lib/events.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+// event than it is to define those that are.
+const NOT_FLOW_EVENTS = new Set([
+  'account.deleted',
+  'account.reset',
+  'device.created',
+  'device.deleted',
+  'device.updated'
+])
+
+const NOT_FLOW_EVENTS_IF_SERVICE_IS_CONTENT_SERVER = new Set([
```
Comment (vbudhram) [inline lib/events.js]: Is there a situation where we would ignore from other services? If so, maybe make it `IGNORE_FLOW_EVENT_FROM_SERVICE`, where each object has name and service?

**Comment by rfk** [comment]
> From now on, we need only make calls to request.emitEvent for both event types

I love the simplicity of the new approach, but I have one quibble with the name `emitEvent` - it's quite close to a existing but different thing in node land, the "event emitter".  We might be able to reduce future confusing by calling it something more explicit like `emitMetricsEvent`.

(IIRC, we used to have this problem with a similarly-generically-named `log.event` and we wound up renaming it to something more explicit to reduce confusion)

**Comment by philbooth** [comment]
Crikey, three reviews, I'm honoured! I've implemented all of your feedback @seanmonstar, @rfk and @vbudhram. Merging...

**Pull request merged**